Source/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -0,0 +1,173 @@
+//
+//  BSG_KSMachHeaders.m
+//  Bugsnag
+//
+//  Created by Robin Macharg on 04/05/2020.
+//  Copyright © 2020 Bugsnag. All rights reserved.
+//
+
+#import <mach-o/dyld.h>
+#import <dlfcn.h>
+#import <Foundation/Foundation.h>
+#import "BSG_KSDynamicLinker.h"
+#import "BSG_KSMachHeaders.h"
+
+BSG_Mach_Binary_Images *bsg_get_mach_binary_images() {
+    return &bsg_mach_binary_images;
+}
+
+/**
+ * Store a Mach binary image-excapsulating struct in a dynamic array.
+ * The array doubles on filling-up.  Typical sizes is expected to be in < 1000 (i.e. 2-3 doublings, at app start-up)
+ * This should be called in a threadsafe way; we lock against a simultaneous add and remove.
+ */
+void bsg_add_mach_binary_image(BSG_Mach_Binary_Image_Info element) {
+    
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+    
+    // Expand array if necessary
+    if (bsg_mach_binary_images.used == bsg_mach_binary_images.size) {
+        bsg_mach_binary_images.size *= 2;
+        bsg_mach_binary_images.contents = (BSG_Mach_Binary_Image_Info *)realloc(bsg_mach_binary_images.contents, bsg_mach_binary_images.size * sizeof(BSG_Mach_Binary_Image_Info));
+    }
+    
+    // Store the value, increment the number of used elements
+    bsg_mach_binary_images.contents[bsg_mach_binary_images.used++] = element;
+    
+    os_unfair_lock_unlock(&bsg_mach_binary_images_access_lock);
+}
+
+/**
+ * Binary images can only be loaded at most once.  We can use the (file)name as a key, without needing to compare the
+ * other fields.  Element order is not important; deletion is accomplished by copying the last item into the deleted
+ * position.
+ */
+void bsg_remove_mach_binary_image(const char *element_name) {
+    
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+    
+    for (size_t i=0; i<bsg_mach_binary_images.used; i++) {
+        BSG_Mach_Binary_Image_Info item = bsg_mach_binary_images.contents[i];
+        
+        if (strcmp(element_name, item.name) == 0) {
```

Comment:
Will name definitely always be unique for binary images? `vmAddress` seems like it would definitely always be unique so it might be good to use that here. Perhaps being overly paranoid - if the filename is guaranteed to be different then this would be fine.

Reply:
I think filename _should_ be unique, but can imagine a pathological scenario where libraries had filenames changed prior to loading (and that's an interesting experiment to try); I've changed the code to use vmAddr as suggested since this is internal and once loaded this should be unique.